runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -839,6 +839,51 @@ J9::CodeGenerator::lowerTreeIfNeeded(
          }
       }
 
+   if (node->getOpCode().isCall() &&
+         node->getSymbol()->getMethodSymbol()->isNative() &&
+         self()->comp()->canTransformUnsafeCopyToArrayCopy())
+      {
+      TR::RecognizedMethod rm = node->getSymbol()->castToMethodSymbol()->getRecognizedMethod();
+
+      if ((rm == TR::sun_misc_Unsafe_copyMemory || rm == TR::jdk_internal_misc_Unsafe_copyMemory0) &&
+            performTransformation(self()->comp(), "O^O Call arraycopy instead of Unsafe.copyMemory: %s\n", self()->getDebug()->getName(node)))
+         {
+         TR::Node *src = node->getChild(1);
+         TR::Node *srcOffset = node->getChild(2);
+         TR::Node *dest = node->getChild(3);
+         TR::Node *destOffset = node->getChild(4);
+         TR::Node *len = node->getChild(5);
+
+         if (self()->comp()->target().is32Bit())
+            {
+            srcOffset = TR::Node::create(TR::l2i, 1, srcOffset);
+            destOffset = TR::Node::create(TR::l2i, 1, destOffset);
+            len = TR::Node::create(TR::l2i, 1, len);
+            src = TR::Node::create(TR::aiadd, 2, src, srcOffset);
+            dest = TR::Node::create(TR::aiadd, 2, dest, destOffset);
+            }
+         else
+            {
+            src = TR::Node::create(TR::aladd, 2, src, srcOffset);
+            dest = TR::Node::create(TR::aladd, 2, dest, destOffset);
+            }
+
+         TR::Node *arraycopyNode = TR::Node::createArraycopy(src, dest, len);
+         TR::TreeTop *arrayCopyTT = TR::TreeTop::create(self()->comp(), arraycopyNode, tt->getNextTreeTop(), tt->getPrevTreeTop());
+
+         tt->getPrevTreeTop()->setNextTreeTop(arrayCopyTT);
+         tt->getNextTreeTop()->setPrevTreeTop(arrayCopyTT);
+         tt->getNode()->recursivelyDecReferenceCount();
```

Comment:
I don't think a recursive dec reference count is what you want here.  Creating a new arraycopy node with children will add the children and increment their reference counts.  I think all you want is to decrement the reference count on each child because you are simply moving the children from one tree to another before unlinking the original tree.

Another option to look into is whether the original node could be "recreated" (via `TR::Node::recreate`).  This might be cleaner, and you wouldn't have to worry about unlinking it.

Reply:
> I think all you want is to decrement the reference count on each child because you are simply moving the children from one tree to another before unlinking the original tree.

Won't `recursivelyDecReferenceCount()` this decrease the reference count of the original node, and its children?

> Another option to look into is whether the original node could be "recreated" (via TR::Node::recreate)

We can't use TR::Node::recreate because the children differ from the original node, which it does not support.

Reply:
Unless I'm mistaken, `recursivelyDecReferenceCount()` won't just do the children, but continue to descend.  You should only have to touch the ref counts on the immediate children.